dev-infra/release/publish/release-notes/context.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {COMMIT_TYPES, ReleaseNotesLevel} from '../../../commit-message/config';
+import {CommitFromGitLog} from '../../../commit-message/parse';
+import {GithubConfig} from '../../../utils/config';
+import {ReleaseNotesConfig} from '../../config/index';
+
+
+/** List of types to be included in the release notes. */
+const typesToIncludeInReleaseNotes =
+    Object.values(COMMIT_TYPES)
+        .filter(type => type.releaseNotesLevel === ReleaseNotesLevel.Visible)
+        .map(type => type.name);
+
+/** Data used for context during rendering. */
+export interface RenderContextData {
+  title: string|false;
+  groupOrder?: ReleaseNotesConfig['groupOrder'];
+  hiddenScopes?: ReleaseNotesConfig['hiddenScopes'];
+  date?: Date;
+  commits: CommitFromGitLog[];
+  version: string;
+  github: GithubConfig;
+}
+
+/** Context class used for rendering release notes. */
+export class RenderContext {
+  /** An array of group names in sort order if defined. */
+  private readonly groupOrder = this.data.groupOrder || [];
+  /** An array of scopes to hide from the release entry output.  */
+  private readonly hiddenScopes = this.data.hiddenScopes || [];
+  /** The title of the release, or undefined if no title should be used. */
+  readonly title = this.data.title;
+  /** An array of commits in the release period. */
+  readonly commits = this.data.commits;
+  /** The version of the release. */
+  readonly version = this.data.version;
+  /** The date stamp string for use in the release notes entry. */
+  readonly dateStamp = buildDateStamp(this.data.date);
+
+  constructor(private readonly data: RenderContextData) {}
+
+  /**
+   * Organizes and sorts the commits into groups of commits.
+   *
+   * Groups are sorted either by default `Array.sort` order, or using the provided group order from
+   * the configuration.  Commits are order in the same order within each groups commit list as they
+   * appear in the provided list of commits.
+   * */
+  asCommitGroups(commits: CommitFromGitLog[]) {
+    /** The discovered groups to organize into. */
+    const groups = new Map<string, CommitFromGitLog[]>();
+
+    // Place each commit in the list into its group.
+    commits.forEach(commit => {
+      const key = commit.npmScope ? `${commit.npmScope}/${commit.scope}` : commit.scope;
+      const commits = groups.get(key) || [];
```

Comment:
It would be better for this variable to not shadow the outer `commits` variable. E.g.:

```suggestion
      const groupCommits = groups.get(key) || [];
```